File: hms_tz/nhif/api/insurance_subscription.js
frappe.ui.form.on('Healthcare Insurance Subscription', {
    setup: function (frm) {
        frm.set_query('healthcare_insurance_coverage_plan', function () {
            return {
                filters: {
                    'insurance_company': frm.doc.insurance_company,
                    'is_active': 1
                }
            };
        });
    },
    coverage_plan_card_number: function (frm) {
        frm.fields_dict.coverage_plan_card_number.$input.focusout(function () {
            frm.trigger("get_patient_name");
            if (!frm.doc.insurance_company.includes("NHIF")) {
                if (!frm.doc.coverage_plan_card_number) return

                setTimeout(() => {
                    frappe.show_alert({
                        message: __('<b>Healthcare Insurance Subscription is submitted</b>'),
                        'indicator': 'success'
                    });
                    frm.save("Submit")
                }, 10000);
            }
        });
    },
    insurance_company: function (frm) {
        frm.trigger("get_patient_name");
        frm.set_value("daily_limit", 0);
    },
    get_patient_name: function (frm) {
        if (!frm.doc.insurance_company.includes("NHIF")) return
        if (!frm.doc.coverage_plan_card_number && !frm.doc.national_id) return

        let args = {
            'patient': frm.doc.patient,
            'patient_name': frm.doc.patient_name
        }

        if (frm.doc.coverage_plan_card_number) {
            args.card_no = frm.doc.coverage_plan_card_number
        } else if (frm.doc.national_id) {
            args.national_id = frm.doc.national_id
        }

        args.ref_doctype = frm.doc.doctype;

        if (!frm.is_new()) {
            args.ref_docname = frm.doc.name;
        }

        frappe.show_alert({
            message: __("Getting patient's information from NHIF"),
            indicator: 'green'
        }, 5);

        frappe.call({
            method: "hms_tz.nhif.api.insurance_subscription.check_patient_info",
            args: args,
            freeze: true,
            freeze_message: __('<i class="fa fa-spinner fa-spin fa-4x"></i>'),
        }).then(data => {
            frappe.utils.play_sound("submit");

            if (data.message) {
                if (data.message != frm.doc.patient_name){
                    frm.set_value("patient_name", data.message);
                    frm.save("Submit")
                }
            } else {
                frappe.utils.play_sound("error");
                frappe.show_alert({
                    message: __("Failed to get patient's information"),
                    indicator: 'red'
                }, 10);
            }
        }).on_error(err => {
            frappe.utils.play_sound("error");
            frappe.show_alert({
                message: __("Failed to get patient's information"),
                indicator: 'red'
            }, 10);
        });
    },
});

File: hms_tz/nhif/api/insurance_subscription.py
# -*- coding: utf-8 -*-
# Copyright (c) 2020, Aakvatech and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _

# from frappe import _
from hms_tz.nhif.api.patient import get_nhif_patient_info



def before_insert(doc, method):
    validate_card_no(doc)
    validate_national_id(doc)


def validate(doc, method):
    validate_card_no(doc)
    validate_national_id(doc)


def on_submit(doc, method):
    set_insurance_card_detail_in_patient(doc)


def on_update_after_submit(doc, method):
    if method != "on_submit" and doc.is_active == 1:
        set_insurance_card_detail_in_patient(doc)


def on_cancel(doc, method):
    set_insurance_card_detail_in_patient(doc)


def validate_card_no(doc):
    if not doc.coverage_plan_card_number:
        return
    
    filters = {
        "is_active": 1,
        "docstatus": 1,
        "coverage_plan_card_number": doc.coverage_plan_card_number,
        "name": ["!=", doc.name]
    }
    
    his = frappe.db.get_all("Healthcare Insurance Subscription", filters=filters, fields=["name", "patient_name"])
    if len(his) > 0:
        frappe.throw(
            f"Cardno: <b>{doc.coverage_plan_card_number}</b> used with HIS: {his[0].name} patient: <b>{his[0].patient_name}</b>, Please change Cardno to Proceed"
        )


def validate_national_id(doc):
    if not doc.national_id:
        return
    
    filters = {
        "is_active": 1,
        "docstatus": 1,
        "national_id": doc.national_id,
        "name": ["!=", doc.name]
    }
    
    his = frappe.db.get_all("Healthcare Insurance Subscription", filters=filters, fields=["name", "patient_name"])
    if len(his) > 0:
        frappe.throw(
            f"NationalID: <b>{doc.national_id}</b> used with HIS: {his[0].name} Patient: <b>{his[0].patient_name}</b>, Please change NationalID to Proceed"
        )


def set_insurance_card_detail_in_patient(doc):
    his_list = frappe.get_all(
        "Healthcare Insurance Subscription",
        filters={
            "patient": doc.patient,
            "docstatus": 1,
            "is_active": 1,
        },
        fields=["coverage_plan_card_number"],
        group_by="coverage_plan_card_number",
    )
    str_coverage_plan_card_number = ""
    card_count = 0
    for card in his_list:
        if card.coverage_plan_card_number:
            card_count += 1
            str_coverage_plan_card_number += card.coverage_plan_card_number + ", "

    frappe.db.set_value(
        "Patient",
        doc.patient,
        {
            "insurance_card_detail": str_coverage_plan_card_number[:-2],
            "card_no": doc.coverage_plan_card_number,
        },
    )


@frappe.whitelist()
def check_patient_info(
    patient,
    patient_name,
    card_no=None,
    national_id=None,
    ref_doctype=None,
    ref_docname=None
):
    if not patient or (not card_no and not national_id):
        return
        
    patient_info = get_nhif_patient_info(
        card_no=card_no,
        national_id=national_id,
        ref_doctype=ref_doctype,
        ref_docname=ref_docname,
        check_patient_info_from_his=True,
    )

    if (
        patient_info and 
        patient_name != patient_info.get("FullName")
    ):
        patient_doc = frappe.get_cached_doc("Patient", patient)
        patient_doc.patient_name = patient_info.get("FullName")
        patient_doc.first_name = patient_info.get("FirstName")
        patient_doc.middle_name = patient_info.get("MiddleName")
        patient_doc.last_name = patient_info.get("LastName")
        patient_doc.sex = patient_info.get("Gender")
        patient_doc.dob = patient_info.get("DateOfBirth")
        patient_doc.product_code = patient_info.get("ProductCode")
        patient_doc.membership_no = patient_info.get("membership_no")
        patient_doc.save(ignore_permissions=True)

        return patient_info.get("FullName")
    
    return None

